File: services/worker/remin_worker/execution.py
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime
from enum import StrEnum
from typing import Any
from uuid import UUID

from .protocol import WorkerError, is_uuid_v7


class CheckRuleType(StrEnum):
    SCHEMA = "schema"
    INVARIANT = "invariant"
    EVIDENCE = "evidence"
    PERMISSION = "permission"
    RESOURCE = "resource"
    QUALITY = "quality"


class RiskLevel(StrEnum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"


class CheckpointPolicy(StrEnum):
    ALWAYS = "always"
    ON_SUCCESS = "on_success"
    NONE = "none"


@dataclass(frozen=True, slots=True)
class CheckRule:
    rule_id: str
    rule_type: CheckRuleType
    description: str
    parameters: dict[str, Any] = field(default_factory=dict)
    required: bool = True


@dataclass(frozen=True, slots=True)
class RetryPolicy:
    max_attempts: int
    backoff_ms: int
    backoff_multiplier: int
    retryable_codes: tuple[str, ...] = ()


@dataclass(frozen=True, slots=True)
class ExecutionUnit:
    unit_id: str
    unit_type: str
    input_schema: str
    output_schema: str
    inputs: dict[str, Any]
    preconditions: tuple[CheckRule, ...]
    postconditions: tuple[CheckRule, ...]
    timeout_ms: int
    retry_policy: RetryPolicy
    idempotency_key: str
    risk_level: RiskLevel
    checkpoint_policy: CheckpointPolicy
    fallback_unit_types: tuple[str, ...] = ()

    def validate(self) -> tuple[WorkerError, ...]:
        errors: list[WorkerError] = []
        if not is_uuid_v7(self.unit_id):
            errors.append(_error("SCHEMA_UUID_V7_REQUIRED", "unit_id必须使用UUIDv7"))
        for field_name, value in (
            ("unit_type", self.unit_type),
            ("input_schema", self.input_schema),
            ("output_schema", self.output_schema),
            ("idempotency_key", self.idempotency_key),
        ):
            if not value.strip():
                errors.append(_error("SCHEMA_REQUIRED_FIELD", f"{field_name}不能为空"))
        if not 1 <= self.timeout_ms <= 86_400_000:
            errors.append(_error("SCHEMA_INVALID_TIMEOUT", "timeout_ms必须位于1毫秒到24小时之间"))
        if not 1 <= self.retry_policy.max_attempts <= 10 or self.retry_policy.backoff_multiplier < 1:
            errors.append(_error("SCHEMA_INVALID_RETRY_POLICY", "重试策略超出允许范围"))
        if self.risk_level is RiskLevel.HIGH and self.checkpoint_policy is not CheckpointPolicy.ALWAYS:
            errors.append(_error("SCHEMA_CHECKPOINT_REQUIRED", "高风险执行单元必须始终创建检查点"))
        rule_ids = [rule.rule_id for rule in (*self.preconditions, *self.postconditions)]
        if len(rule_ids) != len(set(rule_ids)):
            errors.append(_error("SCHEMA_DUPLICATE_RULE_ID", "同一执行单元内的检查规则标识不能重复"))
        if self.unit_type in self.fallback_unit_types:
            errors.append(_error("SCHEMA_INVALID_FALLBACK", "执行单元不能把自身类型作为降级路径"))
        return tuple(errors)


class CheckpointStatus(StrEnum):
    PASSED = "passed"
    FAILED = "failed"
    WARNING = "warning"


@dataclass(frozen=True, slots=True)
class ValidationCheckpoint:
    checkpoint_id: str
    job_id: str
    unit_id: str
    checkpoint_type: CheckRuleType
    status: CheckpointStatus
    rules_version: str
    metrics: dict[str, Any]
    error: WorkerError | None
    created_at: datetime
    resume_token: str | None = None

    def validate(self) -> tuple[WorkerError, ...]:
        errors: list[WorkerError] = []
        for field_name, value in (
            ("checkpoint_id", self.checkpoint_id),
            ("job_id", self.job_id),
            ("unit_id", self.unit_id),
        ):
            if not is_uuid_v7(value):
                errors.append(_error("SCHEMA_UUID_V7_REQUIRED", f"{field_name}必须使用UUIDv7"))
        if self.status is CheckpointStatus.FAILED and self.error is None:
            errors.append(_error("SCHEMA_CHECKPOINT_ERROR_REQUIRED", "失败检查点必须包含结构化错误"))
        if self.status is CheckpointStatus.PASSED and self.error is not None:
            errors.append(_error("SCHEMA_CHECKPOINT_ERROR_UNEXPECTED", "通过的检查点不能包含错误"))
        return tuple(errors)


def _error(code: str, message: str) -> WorkerError:
    return WorkerError(code=code, message=message, retryable=False)


def ensure_uuid(value: str) -> UUID:
    """Convert an already validated identifier for persistence adapters."""
    return UUID(value)

File: crates/core/src/execution.rs
use std::collections::HashSet;

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use uuid::{Uuid, Version};

use crate::AppError;

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum CheckRuleType {
    Schema,
    Invariant,
    Evidence,
    Permission,
    Resource,
    Quality,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct CheckRule {
    pub rule_id: String,
    pub rule_type: CheckRuleType,
    pub description: String,
    pub parameters: Value,
    pub required: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct RetryPolicy {
    pub max_attempts: u8,
    pub backoff_ms: u64,
    pub backoff_multiplier: u8,
    pub retryable_codes: Vec<String>,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum RiskLevel {
    Low,
    Medium,
    High,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum CheckpointPolicy {
    Always,
    OnSuccess,
    None,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ExecutionUnit {
    pub unit_id: Uuid,
    pub unit_type: String,
    pub input_schema: String,
    pub output_schema: String,
    pub inputs: Value,
    pub preconditions: Vec<CheckRule>,
    pub postconditions: Vec<CheckRule>,
    pub timeout_ms: u64,
    pub retry_policy: RetryPolicy,
    pub idempotency_key: String,
    pub risk_level: RiskLevel,
    pub checkpoint_policy: CheckpointPolicy,
    pub fallback_unit_types: Vec<String>,
}

impl ExecutionUnit {
    pub fn validate(&self) -> Result<(), AppError> {
        require_uuid_v7(self.unit_id, "unit_id")?;
        require_non_empty(&self.unit_type, "unit_type")?;
        require_non_empty(&self.input_schema, "input_schema")?;
        require_non_empty(&self.output_schema, "output_schema")?;
        require_non_empty(&self.idempotency_key, "idempotency_key")?;
        if self.timeout_ms == 0 || self.timeout_ms > 86_400_000 {
            return Err(schema_error(
                "SCHEMA_INVALID_TIMEOUT",
                "timeout_ms必须位于1毫秒到24小时之间",
            ));
        }
        if self.retry_policy.max_attempts == 0 || self.retry_policy.max_attempts > 10 {
            return Err(schema_error(
                "SCHEMA_INVALID_RETRY_POLICY",
                "max_attempts必须位于1到10之间",
            ));
        }
        if self.retry_policy.backoff_multiplier == 0 {
            return Err(schema_error(
                "SCHEMA_INVALID_RETRY_POLICY",
                "backoff_multiplier必须大于0",
            ));
        }
        if self.risk_level == RiskLevel::High && self.checkpoint_policy != CheckpointPolicy::Always
        {
            return Err(schema_error(
                "SCHEMA_CHECKPOINT_REQUIRED",
                "高风险执行单元必须始终创建检查点",
            ));
        }
        validate_rules(self.preconditions.iter().chain(self.postconditions.iter()))?;
        if self
            .fallback_unit_types
            .iter()
            .any(|fallback| fallback == &self.unit_type)
        {
            return Err(schema_error(
                "SCHEMA_INVALID_FALLBACK",
                "执行单元不能把自身类型作为降级路径",
            ));
        }
        Ok(())
    }
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum CheckpointType {
    Schema,
    Invariant,
    Evidence,
    Permission,
    Resource,
    Quality,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum CheckpointStatus {
    Passed,
    Failed,
    Warning,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ValidationCheckpoint {
    pub checkpoint_id: Uuid,
    pub job_id: Uuid,
    pub unit_id: Uuid,
    pub checkpoint_type: CheckpointType,
    pub status: CheckpointStatus,
    pub rules_version: String,
    pub metrics: Value,
    pub error: Option<AppError>,
    pub created_at: DateTime<Utc>,
    pub resume_token: Option<String>,
}

impl ValidationCheckpoint {
    pub fn validate(&self) -> Result<(), AppError> {
        require_uuid_v7(self.checkpoint_id, "checkpoint_id")?;
        require_uuid_v7(self.job_id, "job_id")?;
        require_uuid_v7(self.unit_id, "unit_id")?;
        require_non_empty(&self.rules_version, "rules_version")?;
        if self.status == CheckpointStatus::Failed && self.error.is_none() {
            return Err(schema_error(
                "SCHEMA_CHECKPOINT_ERROR_REQUIRED",
                "失败检查点必须包含结构化错误",
            ));
        }
        if self.status == CheckpointStatus::Passed && self.error.is_some() {
            return Err(schema_error(
                "SCHEMA_CHECKPOINT_ERROR_UNEXPECTED",
                "通过的检查点不能包含错误",
            ));
        }
        Ok(())
    }
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum CandidateStatus {
    Pending,
    Running,
    Valid,
    Rejected,
    Selected,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ExplorationCandidate {
    pub candidate_id: Uuid,
    pub job_id: Uuid,
    pub strategy: String,
    pub status: CandidateStatus,
    pub result_ref: Option<String>,
    pub quality_score: Option<f32>,
    pub evidence_score: Option<f32>,
    pub latency_ms: Option<u64>,
    pub resource_cost: Option<f32>,
    pub rejection_reasons: Vec<String>,
}

impl ExplorationCandidate {
    pub fn validate(&self) -> Result<(), AppError> {
        require_uuid_v7(self.candidate_id, "candidate_id")?;
        require_uuid_v7(self.job_id, "job_id")?;
        require_non_empty(&self.strategy, "strategy")?;
        for score in [self.quality_score, self.evidence_score]
            .into_iter()
            .flatten()
        {
            if !(0.0..=1.0).contains(&score) {
                return Err(schema_error(
                    "SCHEMA_INVALID_SCORE",
                    "候选质量与证据分数必须位于0到1之间",
                ));
            }
        }
        if self.status == CandidateStatus::Selected && self.result_ref.is_none() {
            return Err(schema_error(
                "SCHEMA_CANDIDATE_RESULT_REQUIRED",
                "选中的候选路径必须引用已验证结果",
            ));
        }
        if self.status == CandidateStatus::Rejected && self.rejection_reasons.is_empty() {
            return Err(schema_error(
                "SCHEMA_REJECTION_REASON_REQUIRED",
                "被拒绝的候选路径必须记录原因",
            ));
        }
        Ok(())
    }
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum DegradationLevel {
    Full,
    Balanced,
    Core,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct DegradationState {
    pub level: DegradationLevel,
    pub triggers: Vec<String>,
    pub disabled_features: Vec<String>,
    pub entered_at: Option<DateTime<Utc>>,
    pub recover_after: Option<DateTime<Utc>>,
    pub manual_override: bool,
}

impl DegradationState {
    pub fn full() -> Self {
        Self {
            level: DegradationLevel::Full,
            triggers: Vec::new(),
            disabled_features: Vec::new(),
            entered_at: None,
            recover_after: None,
            manual_override: false,
        }
    }

    pub fn validate(&self) -> Result<(), AppError> {
        if self.level == DegradationLevel::Full
            && (!self.triggers.is_empty() || !self.disabled_features.is_empty())
        {
            return Err(schema_error(
                "SCHEMA_INVALID_DEGRADATION_STATE",
                "full状态不能保留降级触发原因或禁用能力",
            ));
        }
        if self.level != DegradationLevel::Full
            && (self.triggers.is_empty() || self.entered_at.is_none())
        {
            return Err(schema_error(
                "SCHEMA_INVALID_DEGRADATION_STATE",
                "降级状态必须记录触发原因和进入时间",
            ));
        }
        if let (Some(entered_at), Some(recover_after)) = (self.entered_at, self.recover_after)
            && recover_after < entered_at
        {
            return Err(schema_error(
                "SCHEMA_INVALID_RECOVERY_TIME",
                "恢复检查时间不能早于降级进入时间",
            ));
        }
        Ok(())
    }

    pub fn can_transition_to(&self, next: DegradationLevel) -> bool {
        self.level == next
            || matches!(
                (self.level, next),
                (DegradationLevel::Full, DegradationLevel::Balanced)
                    | (DegradationLevel::Balanced, DegradationLevel::Full)
                    | (DegradationLevel::Balanced, DegradationLevel::Core)
                    | (DegradationLevel::Core, DegradationLevel::Balanced)
            )
    }
}

fn validate_rules<'a>(rules: impl IntoIterator<Item = &'a CheckRule>) -> Result<(), AppError> {
    let mut identifiers = HashSet::new();
    for rule in rules {
        require_non_empty(&rule.rule_id, "rule_id")?;
        require_non_empty(&rule.description, "description")?;
        if !identifiers.insert(&rule.rule_id) {
            return Err(schema_error(
                "SCHEMA_DUPLICATE_RULE_ID",
                "同一执行单元内的检查规则标识不能重复",
            ));
        }
    }
    Ok(())
}

fn require_uuid_v7(value: Uuid, field: &str) -> Result<(), AppError> {
    if value.get_version() != Some(Version::SortRand) {
        return Err(schema_error(
            "SCHEMA_UUID_V7_REQUIRED",
            format!("{field}必须使用UUIDv7"),
        ));
    }
    Ok(())
}

fn require_non_empty(value: &str, field: &str) -> Result<(), AppError> {
    if value.trim().is_empty() {
        return Err(schema_error(
            "SCHEMA_REQUIRED_FIELD",
            format!("{field}不能为空"),
        ));
    }
    Ok(())
}

fn schema_error(code: &str, message: impl Into<String>) -> AppError {
    AppError::new(code, message, false)
}

#[cfg(test)]
mod tests {
    use serde_json::json;

    use super::*;

    fn valid_unit() -> ExecutionUnit {
        ExecutionUnit {
            unit_id: Uuid::now_v7(),
            unit_type: "document.probe".into(),
            input_schema: "remin://schema/document-probe-input/v1".into(),
            output_schema: "remin://schema/document-probe-output/v1".into(),
            inputs: json!({"file_id": Uuid::now_v7()}),
            preconditions: vec![CheckRule {
                rule_id: "permission.source_readonly".into(),
                rule_type: CheckRuleType::Permission,
                description: "资料源必须位于已注册根目录且只读".into(),
                parameters: json!({}),
                required: true,
            }],
            postconditions: vec![],
            timeout_ms: 5_000,
            retry_policy: RetryPolicy {
                max_attempts: 2,
                backoff_ms: 250,
                backoff_multiplier: 2,
                retryable_codes: vec!["FILE_BUSY".into()],
            },
            idempotency_key: "document.probe:test".into(),
            risk_level: RiskLevel::Low,
            checkpoint_policy: CheckpointPolicy::OnSuccess,
            fallback_unit_types: vec![],
        }
    }

    #[test]
    fn execution_unit_accepts_a_complete_atomic_contract() {
        valid_unit().validate().expect("valid execution unit");
    }

    #[test]
    fn high_risk_unit_requires_an_always_checkpoint() {
        let mut unit = valid_unit();
        unit.risk_level = RiskLevel::High;
        let error = unit.validate().expect_err("checkpoint must be required");
        assert_eq!(error.code, "SCHEMA_CHECKPOINT_REQUIRED");
    }

    #[test]
    fn job_and_unit_identifiers_must_be_uuid_v7() {
        let mut unit = valid_unit();
        unit.unit_id = Uuid::nil();
        let error = unit.validate().expect_err("uuid version must be checked");
        assert_eq!(error.code, "SCHEMA_UUID_V7_REQUIRED");
    }

    #[test]
    fn degradation_changes_one_level_at_a_time() {
        let full = DegradationState::full();
        assert!(full.can_transition_to(DegradationLevel::Balanced));
        assert!(!full.can_transition_to(DegradationLevel::Core));
    }
}
